### DISCUSSION TITLE
Function input for link spring coefficient

### DISCUSSION THREAD

**Discussion by daniel-ji** [cosmograph-org/cosmos#55] [Ideas]
Hi,

I'm currently working on a project for visualizing the clusters of a molecular dataset and I've been using this library, which is absolute fantastic! I recently opened (and then closed) an issue regarding adjusting the link spring coefficient for every given node. For my use case, I wanted to adjust the link spring coefficient so that larger clusters would appear bigger, since right now the spring coefficient is fixed and the larger clusters just become more dense instead of becoming larger. For my use case, a possible solution I was thinking of was that instead of passing in a number for the spring coefficient, a function could be passed in instead (with the link as the parameter) so that when the source and target nodes of the links have a large number of adjacent nodes, I could return a lower link spring coefficient value accordingly and effectively enlarge the cluster. 

I'm a university student and I would love to contribute to this library, so I'm willing to help build the aforementioned feature if it is feasible and keeps the graph visualization high-performing and efficient. 

Thank you!

Best,
Daniel

**Comment by rokotyan** [comment]
Hi @daniel-ji, we welcome any contributions and actually have that feature in our to do list (https://github.com/cosmograph-org/cosmos/issues/14)!

So if you have time and motivation to work on the feature, please go ahead! We'll try to provide necessary support. 
Right now the `linkSpring` coefficient is passed as a constant value to the corresponding shader: https://github.com/cosmograph-org/cosmos/blob/f0c4bbd6204dc6c3c3d7aeb4e1df348589d48121/src/modules/ForceLink/force-spring.ts#L8

What you'll need to do instead, is to initialize a texture for it and add it to the shader.

**Comment by daniel-ji** [comment]
Sounds good, I'll take a look. Thank you!

Best,
Daniel
